File: leetcode-go/Q051-Q100/Q098_Validate_Binary_Search_Tree/Q098_Validate_Binary_Search_Tree.go
package Q098_Validate_Binary_Search_Tree

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	return isValid(root, math.MinInt64, math.MaxInt64)
}

func isValid(root *TreeNode, min, max int64) bool {
	if root == nil {
		return true
	}
	if int64(root.Val) <= min || int64(root.Val) >= max {
		return false
	}

	return isValid(root.Left, min, int64(root.Val)) && isValid(root.Right, int64(root.Val), max)
}

File: src/Y2018/tree/Q098_Validate_Binary_Search_Tree.java
package Y2018.tree;

public class Q098_Validate_Binary_Search_Tree {

    public boolean isValidBST(TreeNode root) {
        return isValid(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean isValid(TreeNode root, long lowerBound, long upperBound) {
        if (root == null) {
            return true;
        }
        if (root.val <= lowerBound || root.val >= upperBound) {
            return false;
        }
        return isValid(root.left, lowerBound, root.val) && isValid(root.right, root.val, upperBound);
    }

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

}
